p2p: Begin encapsulation

theuni opened the pull request:
This work creates CConnman. The idea is to begin moving data structures and functionality out of globals in net.h and into an instanced class, in order to avoid side-effects in networking code. Eventually, an (internal) api begins to emerge, and as long as the conditions of that api are met, the inner-workings may be a black box.

For now (for ease), a single global CConnman is created. Down the road, the instance could be passed around instead. Also, CConnman should be moved out of net.h/net.cpp, but that can be done in a separate move-only-ish step.

A few todos remain here:
- ~~there are `if(g_connman)`'s everywhere. A few places should assume that a connman is running (or it should be passed around). For example, ProcessMessages.~~
- ~~GetRelayTx/RelayTransaction are awkward. Where should mapRelay live? Should it be dealt with directly rather than hidden inside of relay functions? (rebased on top of #8082)~~
- ~~How to deal with CConnman parameters? Continue adding them to Start(), or have it parse the program options itself? (Created CConnman::Options)~~
- ~~RPC needs a new error for dealing with the "p2p not enabled" case.~~

Todos for a follow-up PR:
- Actually create connman.h/cpp and move the class there.
- Pass CConnman into QT similar to the wallet
- ~~Drop StartNode/StopNode and just use CConnman::Stop/Start directly (Need to make sure that MapPort can be moved directly into Init without side-effects)~~
- Pass CChainParams into CConnman and pass it around
- ~~Create CConnmanOptions to reduce the verbosity of CConnman::Start()~~
- Deal with fDiscover/fListen/etc in CConnman

theuni committed:
net: move CBanDB and CAddrDB out of net.h/cpp

This will eventually solve a circular dependency

theuni committed:
net: Create CConnman to encapsulate p2p connections

theuni committed:
net: Move socket binding into CConnman

theuni committed:
net: move OpenNetworkConnection into CConnman

theuni committed:
net: move ban and addrman functions into CConnman

theuni committed:
net: Add oneshot functions to CConnman

theuni committed:
net: move added node functions to CConnman

theuni committed:
net: Add most functions needed for vNodes to CConnman

theuni committed:
net: create generic functor accessors and move vNodes to CConnman

theuni committed:
net: move whitelist functions into CConnman

theuni committed:
net: move nLastNodeId to CConnman

theuni committed:
net: move nLocalHostNonce to CConnman

This behavior seems to have been quite racy and broken.

Move nLocalHostNonce into CNode, and check received nonces against all
non-fully-connected nodes. If there's a match, assume we've connected
to ourself.

theuni committed:
net: move messageHandlerCondition to CConnman

theuni committed:
net: move send/recv statistics to CConnman

theuni committed:
net: move SendBufferSize/ReceiveFloodSize to CConnman

theuni committed:
net: move nLocalServices/nRelevantServices to CConnman

These are in-turn passed to CNode at connection time. This allows us to offer
different services to different peers (or test the effects of doing so).

theuni committed:
net: move semOutbound to CConnman

theuni committed:
net: SocketSendData returns written size

theuni committed:
net: move max/max-outbound to CConnman

NicolasDorier commented:
nit typo: "SocketSendData re*s*urns written size"

sipa commented on src/net.h:
```diff
@@ -490,12 +490,17 @@ class CNode
     CNode(const CNode&);
     void operator=(const CNode&);
 
+    uint64_t nLocalHostNonce;
```
This won't work. It needs to be a ConnMan-scoped variable, not a CNode-scoped one, as the nonce will come back through another link than the one we sent it through.

theuni replied:
@sipa See https://github.com/bitcoin/bitcoin/pull/8085/commits/46cf6754cf6ac7bab76fb61ad9494741de235f6c#diff-9a82240fe7dfe86564178691cc57f2f1R351.

I started by simply moving this to CConnman as you suggest, but it's racy and dependent on the optimistic send succeeding (nonce is set in one thread, tested in another). With this change, I believe we now guard against rapid-fire connections that may connect to self, which would have been missed before.

sipa replied:
Nevermind, indeed.

sipa commented:
#8082 moves mapRelay from net to main. I think it indeed belongs with node processing code and not in base networking code.

theuni commented:
@sipa Indeed, that's definitely the correct approach and the change here can be dropped.

theuni committed:
net: Pass best block known height into CConnman

CConnman then passes the current best height into CNode at creation time.

This way CConnman/CNode have no dependency on main for height, and the signals
only move in one direction.

This also helps to prevent identity leakage a tiny bit. Before this change, an
attacker could theoretically make 2 connections on different interfaces. They
would connect fully on one, and only establish the initial connection on the
other. Once they receive a new block, they would relay it to your first
connection, and immediately commence the version handshake on the second. Since
the new block height is reflected immediately, they could attempt to learn
whether the two connections were correlated.

This is, of course, incredibly unlikely to work due to the small timings
involved and receipt from other senders. But it doesn't hurt to lock-in
nBestHeight at the time of connection, rather than letting the remote choose
the time.

theuni committed:
net: handle nodesignals in CConnman

theuni committed:
net: Pass CConnection to wallet rather than using the global

arowser commented:
Can one of the admins verify this patch?

theuni committed:
net: pass CClientUIInterface into CConnman

theuni commented:
@laanwj I pushed a bunch of new commits to address the review you gave me. The commit history doesn't really make sense now, though, as there's a good bit of reverting. Please let me know if you'd like me to squash down to a more logical history for easier review, or continue to pile up changes and squash at the end.

theuni committed:
net: Add rpc error for missing/disabled p2p functionality

theuni committed:
net: Pass CConnman around as needed

theuni committed:
net: Drop StartNode/StopNode and use CConnman directly

theuni committed:
net: Introduce CConnection::Options to avoid passing so many params

theuni committed:
net: add nSendBufferMaxSize/nReceiveFloodSize to CConnection::Options

theuni committed:
net: move vNodesDisconnected into CConnman

theuni commented:
Rebased to master and squashed down changes so that the history makes sense again.

Though there's still a good bit left to do, I think this is a good stopping point. I'll avoid making further changes here unless requested.

theuni commented:
Rebased to master after #8082, and dropped that commit here. Should be ready for review.

kazcw commented on src/init.cpp:
```diff
@@ -69,6 +70,7 @@ static const bool DEFAULT_REST_ENABLE = false;
 static const bool DEFAULT_DISABLE_SAFEMODE = false;
 static const bool DEFAULT_STOPAFTERBLOCKIMPORT = false;
 
+std::shared_ptr<CConnman> g_connman;
```
I don't see g_connman requiring any refcounting, could this be a plain unique_ptr?

theuni replied:
This was originally passed around differently, but refactored to essentially be a unique_ptr. So yes :)

Though it's worth mentioning that the global is only temporary until the rpc/qt interfaces learn how to accept/maintain an instance. So it's probably even more important to not encourage any sharing semantics.

Will change.

kazcw commented on src/net.cpp:
```diff
@@ -1945,9 +1919,45 @@ void static Discover(boost::thread_group& threadGroup)
+    return nLastNodeId.fetch_add(1, std::memory_order_relaxed);
+}
+
+bool CConnman::Start(boost::thread_group& threadGroup, CScheduler& scheduler, std::string& strNodeError, Options connOptions)
```
Lifetime safety is more robust without the existence of constructed-not-started and stopped-not-destructed twilight states. What do you think about doing the Start/Stop stuff in the ctor/dtor -- would the stuff that AppInit2 currently does between construction and Start make sense to put into the CConnman::Options?

theuni replied:
The reason for start/stop doesn't really become obvious until the async behavior comes in. As a next step, all of the connecting/resolving behavior will be replaced with async events handled via callbacks. It's necessary to be able to have some basic connman functionality before the event thread spins up (or in case it's never going to). The most straightforward example is that we'll want to stop app startup if all bind()s fail, before spawning any connection threads.

As a better example, you can see that none of the tests require that Start() has been called.

kazcw replied:
I'd be more comfortable with a builder doing those binds or something -- it just seems like it would be easy for a future change to introduce an unchecked dependency on the CConnman state, that could make for subtle and unpredictable bugs. But this is a complex change you're certainly more familiar with than I, so ACK however you decide to ensure safety against that kind of case.

theuni replied:
Roger. I see your point and agree, I'm just avoid overthinking this until the async stuff is added. Once it's all hooked up, maybe we can switch to a model where Start() returns some kind of object that guarantees a certain state that CConnman does not.

The alternative would be to move out things that don't require a running state (addrman/bandb as examples) and let them have separate life-cycles, but I'd like to do that as a next step once all callers understand that those things aren't global.

kazcw commented on src/net.h:
```diff
@@ -75,24 +80,218 @@ static const size_t DEFAULT_MAXSENDBUFFER    = 1 * 1000;
+    bool OpenNetworkConnection(const CAddress& addrConnect, CSemaphoreGrant *grantOutbound = NULL, const char *strDest = NULL, bool fOneShot = false);
+    bool CheckIncomingNonce(uint64_t nonce);
+
+    bool ForEachNode(std::function<bool(CNode* pnode)> func);
```
Nit: std::function indirection is fairly heavyweight, templating these would make the functors inlineable.

theuni replied:
Good point.

kazcw commented on src/main.cpp:
```diff
@@ -2927,7 +2920,7 @@ static void NotifyHeaderTip() {
  * or an activated best chain. pblock is either NULL or a pointer to a block
  * that is already loaded (to avoid loading it again from disk).
  */
```
Nit: could use a comment documenting connman nullability.

theuni replied:
Agreed, will do. In case it's not clear, the intention is to allow us to avoid skipping the network altogether for utility functions. Beyond the startup stuff where it's currently skipped, in the future it may be interesting to do things like on-the-fly reindexing without having to shut down.

kazcw commented on src/main.cpp:
```diff
@@ -4418,17 +4412,60 @@ bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+    // Use deterministic randomness to send to the same nodes for 24 hours
+    // at a time so the addrKnowns of the chosen nodes prevent repeats
+    static uint64_t salt0 = 0, salt1 = 0;
+    while (salt0 == 0 && salt1 == 0) {
```
There's a newer version of this initialization in master from #8173, the commit is 8884830.

theuni replied:
Thanks, I'll note this for the next rebase round.

kazcw commented:
utACK the src/*.* changes at f86d0be (with minor notes in comments) [didn't review the subdirectory stuff thoroughly; I'm less familiar with that code]

theuni commented:
@kazcw Thanks for the excellent review! I'll pick this back up in ~2 weeks. Obviously if that's going to stall progress on lock refactors too much, go ahead with them and I'll cope with it here.

JeremyRubin commented on src/addrdb.cpp:
```diff
@@ -0,0 +1,218 @@
+    // Generate random temporary filename
+    unsigned short randv = 0;
+    GetRandBytes((unsigned char*)&randv, sizeof(randv));
+    std::string tmpfn = strprintf("banlist.dat.%04x", randv);
```
4 digits isn't enough to prevent collisions here... I think if this is being used for some logging purpose it would be better to use either/both date or hash of file or if it isn't then use a deterministic name.

JeremyRubin commented on src/addrdb.cpp:
```diff
@@ -0,0 +1,218 @@
+    ssBanlist << FLATDATA(Params().MessageStart());
+    ssBanlist << banSet;
+    uint256 hash = Hash(ssBanlist.begin(), ssBanlist.end());
+    ssBanlist << hash;
```
Maybe best to break out the serialization/deserialization from the IO, will be better to test later.

JeremyRubin commented on src/addrdb.cpp:
```diff
@@ -0,0 +1,218 @@
+
+CBanDB::CBanDB()
+{
+    pathBanlist = GetDataDir() / "banlist.dat";
```
Perhaps un-hardcode "banlist.dat"

JeremyRubin commented on src/net.h:
```diff
@@ -96,9 +96,18 @@ class CConnman
+        int nMaxConnections = 0;
+        int nMaxOutbound = 0;
+        int nBestHeight = 0;
+        CClientUIInterface* interface = nullptr;
```
nit: unclear why interface is in options when it doesn't really seem to be an option. Perhaps having 
struct Context

JeremyRubin replied:
nit: unclear why interface is in options when it doesn't really seem to be an option. Perhaps having a struct Context to hold state such as interface, scheduler, and threadgroup would be a better encapsulation.

theuni replied:
This will eventually hold pointers to several structures, chainparams for
example.

It may be reasonable to split it into a set of truly immutable options, and
stateful app pointers.

kazcw replied:
ChainParams seems like it could qualify as an option. In the long term I think it would make more sense for the UI to connect to the various backend signals rather than the net layer having any knowledge of the UI, but that change seems outside the scope of this PR.

JeremyRubin commented on src/rpc/net.cpp:
```diff
@@ -303,23 +295,21 @@ UniValue getaddednodeinfo(const UniValue& params, bool fHelp)
             {
-                laddedNodes.push_back(strAddNode);
+                laddedNodes.assign(1, strAddNode);
                 break;
```
note: This section is a bit more than just a code move, but it seems to have no functional change.

JeremyRubin commented on src/net.cpp:
```diff
@@ -2531,6 +2521,50 @@ void CNode::EndMessage(const char* pszCommand) UNLOCK_FUNCTION(cs_vSend)
     LEAVE_CRITICAL_SECTION(cs_vSend);
 }
 
+bool CConnman::ForEachNode(std::function<bool(CNode* pnode)> func)
```
Nit: I would rename `ForEachNode` as `ForEachNodeContinueIf` or something to signify that the return value being true is critical. I would then make a version of `ForEachNode` that does not check for the return value.

- This is more consistent* with other "foreach" API's generally, eg, boost foreach or javascript foreach iterate over all elements
 * although boost foreach does allow break
- It is unclear if `ForEachNodeContinueIf` is/will be used anywhere in the codebase, but it could introduce bugs (say, if false is somehow returned by buggy code elsewhere or later), so it's probably better to not have a check that is never used.

theuni replied:
Ack. I meant to document these, but self-doc would be better.

The intention is to use these internally later, for example to replace the
FindNode's which iterate and early-return when possible, to avoid holding
the lock longer than necessary.

But since those changes aren't part of this pull, I think that simply
dropping the bool return and reintroducing it as ForEachNodeContinueIf
later is the way to go.

JeremyRubin replied:
I'll take care of it...

JeremyRubin commented on src/wallet/wallet.cpp:
```diff
@@ -1342,8 +1342,15 @@ bool CWalletTx::RelayWalletTransaction(CConnman* connman)
-            return true;
+            if (connman) {
+                CInv inv(MSG_TX, GetHash());
+                connman->ForEachNode([&inv](CNode* pnode)
```
Nit: appears correct, but isn't this the same code as in RelayTransaction?

theuni replied:
Yes, but it's simple enough that I didn't bother factoring it out. No
reason not to, except that there's not an obvious place to stick it.
On Jun 14, 2016 19:26, "Jeremy Rubin" <notifications@github.com> wrote:

> In src/wallet/wallet.cpp
> <https://github.com/bitcoin/bitcoin/pull/8085#discussion_r67076355>:
>
> > @@ -1342,8 +1342,15 @@ bool CWalletTx::RelayWalletTransaction(CConnman* connman)
> >      {
> >          if (GetDepthInMainChain() == 0 && !isAbandoned() && InMempool()) {
> >              LogPrintf("Relaying wtx %s\n", GetHash().ToString());
> > -            RelayTransaction((CTransaction)*this);
> > -            return true;
> > +            if (connman) {
> > +                CInv inv(MSG_TX, GetHash());
> > +                connman->ForEachNode([&inv](CNode* pnode)
>
> Nit: appears correct, but isn't this the same code as in RelayTransaction?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/8085/files/a57b3f21e3228f838382e7510dc4e8f2bd023237..a01fa6c5a2e5884af64c0153dbfbd6a848128129#r67076355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAZdE-sEqwxHrvI3wdUKtoHlUCgD1RM2ks5qLzi2gaJpZM4Ij8mn>
> .
>

JeremyRubin commented on src/net.cpp:
```diff
@@ -2337,6 +2338,16 @@ uint64_t CConnman::GetLocalServices() const
 
+void CConnman::SetBestHeight(int height)
+{
+    nBestHeight.store(height, std::memory_order_release);
```
I'm not sure this actually presents a bug, but it might at least exhibit some undesired behavior.

This should replace the value only if it is greater than the previous value, unless it is desirable to roll back (but this seems like a different case?).


This may also not play well with out of order headers.

In any case, I'm not sure this is a consensus critical thing, but it could be the source of future bugs or further identity leaks.

This (might) also be fixed by moving the call to SetBestHeight inside of `Lock(cs_main)` in `ActivateBestChain`. Would also be reasonable to just document (or self-doc) it a bit better about what you are expecting this to do.

sipa replied:
I don't think this matters much.

JeremyRubin commented on src/net.cpp:
```diff
@@ -1976,7 +1982,7 @@ bool CConnman::Start(boost::thread_group& threadGroup, CScheduler& scheduler, ui
     }
 
-    uiInterface.InitMessage(_("Loading banlist..."));
+    clientInterface->InitMessage(_("Loading banlist..."));
```
may want to null check...

JeremyRubin commented on src/net.cpp:
```diff
@@ -1962,7 +1967,8 @@ bool CConnman::Start(boost::thread_group& threadGroup, CScheduler& scheduler, ui
 
-    uiInterface.InitMessage(_("Loading addresses..."));
+    clientInterface = interfaceIn;
+    clientInterface->InitMessage(_("Loading addresses..."));
```
may want to null check...

theuni replied:
Yep, thanks.

JeremyRubin commented on src/init.cpp:
```diff
@@ -1423,9 +1424,14 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
+    Discover(threadGroup);
+
+    // Map ports with UPnP
+    MapPort(GetBoolArg("-upnp", DEFAULT_UPNP));
```
nit: I think I would separate out the GetBoolArg put it into Options.

Then keep the calls to MapPort into Start/Stop.

seems a little more pragmatic & good for debugging (e.g., can dump a nodes options and know how the node was configured)

theuni replied:
Yes, I plan to move the remaining args into Options as a next step. There are a handful left to go still.

JeremyRubin committed:
Made the ForEachNode* functions in src/net.cpp more pragmatic and self documenting

instagibbs commented on src/rpc/misc.cpp:
```diff
@@ -89,7 +89,8 @@ UniValue getinfo(const UniValue& params, bool fHelp)
     obj.push_back(Pair("timeoffset",    GetTimeOffset()));
-    obj.push_back(Pair("connections",   (int)vNodes.size()));
+    if(g_connman)
+        obj.push_back(Pair("connections",   (int)g_connman->GetNodeCount(CConnman::CONNECTIONS_ALL)));
```
nit: perhaps just add 0 if !g_connman or some relevant message

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -635,7 +638,7 @@ UniValue submitblock(const UniValue& params, bool fHelp)
     submitblock_StateCatcher sc(block.GetHash());
     RegisterValidationInterface(&sc);
-    bool fAccepted = ProcessNewBlock(state, Params(), NULL, &block, true, NULL);
+    bool fAccepted = ProcessNewBlock(state, Params(), NULL, &block, true, NULL, g_connman.get());
```
check for !g_connman

kazcw commented on src/net.cpp:
```diff
@@ -2185,18 +2305,33 @@ uint64_t CNode::GetOutboundTargetBytesLeft()
+    return nLocalServices;
+}
+
+void CConnman::SetBestHeight(int height)
```
nit: I think this could use documentation that the setter requires cs_main to be held to prevent older values from overwriting newer values. Also, I don't see any other data that would need to be synchronized with the height via acquire/release semantics; if there are no writes before the release that need to be visible side-effects after the acquire, relaxed load/store would be clearer.

JeremyRubin replied:
@kazcw see https://github.com/bitcoin/bitcoin/pull/8085#discussion_r67177507.

I don't even think holding cs_main would help unless the new height is also checked to be > than the old one (suppose an orphaned block gets relayed, this would decrease the height).

kazcw replied:
Oh, I see... that issue is not introduced by this PR; what's called GetHeight in main and is not not expected to be non-decreasing is used in net as a "best height". Ensuring modification order for nBestHeight matches the order established by cs_main would be sufficient to maintain the current behaviour.

It does seem like it would be better to have a high-water-mark variable protected by cs_main that main code could use to signal when a new *best* height was reached. I don't see the need for that change necessarily to be included in this PR though.

JeremyRubin replied:
I would agree that this PR doesn't need any behavior change. Perhaps just renaming BestHeight -> LastSeenHeight or something. Also a proper best height is semantically unclear -- is it MaxHeight or MaxWorkHeight

theuni replied:
Yea, I don't believe this deviates from the current behavior much, other than slightly obfuscating the exact current known height, since the value is grabbed at connection time now rather than during the handshake.

Though there's the advantage of using the accurate cached value based on a push event, rather than requiring a lock-and-pull as before. I don't think there's any need to try to replicate that old behavior here.

JeremyRubin committed:
Convert ForEachNode* functions to take a templated function argument rather than a std::function to eliminate std::function overhead

theuni commented:
Addressed most (all?) of the comments here, and added some commits.

For future comparison, the current tip is https://github.com/theuni/bitcoin/commit/f96d164f83239e74288ea6a707228be9a4f3ae27. I'll now work on rebasing to master.

theuni committed:
gui: add NodeID to the peer table

btcdrak commented:
needs rebase.

laanwj commented:
Yes, needs rebase, should be merged not very long after that IMO

theuni commented:
I have this rebased locally, I've been pulling out more controversial chunks for individual PR. Would you prefer to just do it all in one go here?

btcdrak commented:
@theuni all in one go please.

theuni commented:
ok, will push up a rebased version tomorrow.

theuni committed:
net: move MAX_FEELER_CONNECTIONS into connman

theuni commented:
Rebased. This has been rebased through lots of changes, though I believe the newer stuff is accounted for.

I'd recommend taking a look at the final CConnman first to get an idea of how it looks, then each individual commit should make more sense.

@jonasschnelli 1cf69d418576a0096f8f3e83494a107774dde2ed is included as we discussed, so that kicking can be done in a more straightforward way, but it's also just nice to have in the gui.

Because this is such a monster, at this point I'd prefer to add fixes on top rather than rebasing to include them.

jonasschnelli commented:
Started IBDing a node with this PR (`-listen=0`, connected to a in-sync mainnet node on the same system). Its built with `--enable-debug` and IBDs with `-debug=net`, `-prune=550` and `-dbcache=4000`.

Reached progress `progress=0.097296` after 4h where I normally do a full IBD (with `--enable-debug`)  under 4h on the same machine.

Either the `-debug=net` is causing the slow down or something regarding this PR.
I can try to run more authentic benchmarks though (direct compare master against this PR on a "quite" machine).

theuni commented:
@jonasschnelli Thanks for having a look. I'll try to reproduce.

laanwj commented:
> Either the -debug=net is causing the slow down 

Hah, I wouldn't be surprised.

Is this performance regression serious enough to block this, or shouldit be solved later? This is a large change, which has had a lot of review by now, and it again needs rebase...

jonasschnelli commented:
Sorry this was my fault in doing benchmark in conjunctions with `--enable-debug`

laanwj commented:
> Sorry this was my fault in doing benchmark in conjunctions with --enable-debug

Ok, good to know

sipa commented on src/net.h:
```diff
@@ -95,28 +95,32 @@ CNode* FindNode(const CService& ip);
+    bool BindListenPort(const CService &bindAddr, std::string& strError, bool fWhitelisted = false);
+
 private:
+    struct ListenSocket {
```
Just a tip, and not a suggestion to change in this PR right now, you can just say 'struct ListenSocket;' here, and fully define it in the cpp file as 'struct CConnman::ListenSocket { ... };', since I don't think any of the users of this class need the fully materialized definition.

theuni replied:
Hmm, I've never had luck with forward-declaring nested classes. I'll give it a shot!

sipa commented on src/net.h:
```diff
@@ -347,12 +393,6 @@ class CNode
 
-    // Denial-of-service detection/prevention
-    // Key is IP address, value is banned-until-time
-    static banmap_t setBanned;
```
I am *so* happy to see these misplaced static fields disappear!

sipa commented on src/rpc/net.cpp:
```diff
@@ -489,6 +489,8 @@ UniValue setban(const UniValue& params, bool fHelp)
                             + HelpExampleCli("setban", "\"192.168.0.0/24\" \"add\"")
                             + HelpExampleRpc("setban", "\"192.168.0.6\", \"add\", 86400")
                             );
+    if(!g_connman)
```
Style nit: space after if.

sipa commented on src/main.cpp:
```diff
@@ -3081,15 +3082,15 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
-                    LOCK(cs_vNodes);
-                    BOOST_FOREACH(CNode* pnode, vNodes) {
+                if(connman) {
+                    connman->ForEachNode([nNewHeight, nBlockEstimate, &vHashes](CNode* pnode) {
```
Not being very familiar with c++11 lambdas yet, what is the reason for explicitly listing the captures, rather than using [&] or [=] ?

sipa commented on src/main.cpp:
```diff
@@ -484,11 +484,12 @@ void MaybeSetPeerAsAnnouncingHeaderAndIDs(const CNodeState* nodestate, CNode* pf
             // blocks using compact encodings.
-            CNode* pnodeStop = FindNode(lNodesAnnouncingHeaderAndIDs.front());
-            if (pnodeStop) {
+            bool found = connman.ForNode(lNodesAnnouncingHeaderAndIDs.front(), [fAnnounceUsingCMPCTBLOCK, nCMPCTBLOCKVersion](CNode* pnodeStop){
```
Not being very familiar with c++11 lambdas yet, what is the reason for explicitly listing the captures, rather than using [=] ?

theuni replied:
Just personal preference, I like being explicit about what the lambda is changing. In this case, = is probably desirable, it's [&] that I try to avoid.

sipa replied:
Seems reasonable.

sipa replied:
It seems you can use [=,x&,y&,...] as well to mean "capture everything by value, but x and y by reference", neat.

theuni replied:
Ah, perfect. That's great as a general rule.

sipa commented on src/net.cpp:
```diff
@@ -2671,6 +2659,63 @@ void CNode::EndMessage(const char* pszCommand) UNLOCK_FUNCTION(cs_vSend)
+{
+    bool ret = true;
+    LOCK(cs_vNodes);
+    for (auto&& node : vNodes)
```
A universal reference rvalue seems unnecessary here, since they're just pointers you can copy.

theuni replied:
Yea, this is just a habit I've picked up. See the "notes" here: http://en.cppreference.com/w/cpp/language/range-for

sipa commented on src/net.cpp:
```diff
@@ -2041,9 +2039,13 @@ bool StartNode(CConnman& connman, boost::thread_group& threadGroup, CScheduler&
-bool CConnman::Start(boost::thread_group& threadGroup, CScheduler& scheduler, std::string& strNodeError)
+NodeId CConnman::GetNewNodeId()
 {
+    return nLastNodeId.fetch_add(1, std::memory_order_relaxed);
```
Neat.

sipa commented on src/net.h:
```diff
@@ -109,9 +109,18 @@ class CConnman
     CConnman();
     ~CConnman();
-    bool Start(boost::thread_group& threadGroup, CScheduler& scheduler, ServiceFlags nLocalServicesIn, ServiceFlags nRelevantServicesIn, int nMaxConnectionsIn, int nMaxOutboundIn, int nBestHeightIn, CClientUIInterface* interfaceIn, std::string& strNodeError);
+    bool Start(boost::thread_group& threadGroup, CScheduler& scheduler, std::string& strNodeError, Options options);
```
rvalue reference for options?

theuni replied:
They're passed by value so that they can be std::move'd in if desired, but as-is, all members are fundamental, so they'll all be copied anyway

sipa commented:
Concept and overall code review ACK. I think you still need to address https://github.com/bitcoin/bitcoin/pull/8085/files#r66671033. I also left a number of nits in various commits, but none of them are blockers.

Needs rebase, though.

theuni commented:
@sipa Yikes, thanks for noticing the missed changes. I'll rebase now and make sure that matches.
Thanks for the review!

theuni commented:
Rebased, fixed up the salt changes, and squashed the win32 build fix.

laanwj commented:
Thanks for the review @sipa.

laanwj commented:
ACK 0103c5b

rebroad commented on src/main.cpp:
```diff
@@ -5238,7 +5254,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                         (!IsWitnessEnabled(chainActive.Tip(), chainparams.GetConsensus()) || State(pfrom->GetId())->fHaveWitness)) {
                         inv.type |= nFetchFlags;
-                        if (nodestate->fProvidesHeaderAndIDs && !(nLocalServices & NODE_WITNESS))
+                        if (nodestate->fProvidesHeaderAndIDs && !(pfrom->GetLocalServices() & NODE_WITNESS))
```
I think the current naming is a little confusing as to be a function of pfrom implies that the "Local" Services are local to pfrom. "nLocalServices" seemed more intuitively compatible with talking about the services local to our peer.

rebroad commented on src/main.cpp:
```diff
@@ -4984,7 +5019,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
         // Disconnect if we connected to ourself
-        if (nNonce == nLocalHostNonce && nNonce > 1)
+        if (pfrom->fInbound && !connman.CheckIncomingNonce(nNonce))
```
The code isn't as simple as it used to be - what was wrong with "nNonce == nLocalHostNonce"? it was so much clearer what the code was doing before...

sipa commented:
@rebroad It was wrong, as is explained in the commit. If two connections arrive before the version message cycles back, nLocalHostNonce would be overwritten by the second connection before the comparison occurs.
